File: app/src/main/java/com/anthonysottile/kenken/ui/SquareDrawingDimensions.java
package com.anthonysottile.kenken.ui;

import android.graphics.Canvas;
import android.graphics.Paint;

public class SquareDrawingDimensions {

    private static final int cageTextLeftMargin = 5;

    private static final int cageFontSizeBase = 33;

    private static int getCageTextFontSize(int order) {
        return SquareDrawingDimensions.cageFontSizeBase - 2 * order;
    }

    private static String getTestCandidateString(int order) {
        // If the order is >= 6 then we are returning only half
        // Otherwise we are returning a full candidates string
        StringBuilder sb = new StringBuilder();
        if (order >= 6) {
            boolean first = true;
            // Round up
            int halfOrder = (int) (1.0 * order / 2 + .9);
            for (int i = 0; i < halfOrder; i += 1) {
                if (!first) {
                    sb.append(' ');
                }
                first = false;
                sb.append(i + 1);
            }
        } else {
            boolean first = true;
            for (int i = 0; i < order; i += 1) {
                if (!first) {
                    sb.append(' ');
                }
                first = false;
                sb.append(i + 1);
            }
        }

        return sb.toString();
    }

    private final int squareWidth;
    private final int squareHeight;
    private final int order;

    private final int cageTextFontSize;
    private final int valueTextFontSize;
    private int candidatesTextFontSize;

    private Paint cageTextPaint;
    private Paint valueTextPaint;
    private Paint candidatesTextPaint;


    private int getLeft(int x) {
        return UIConstants.BorderWidth * (x + 1) + x * this.squareWidth;
    }

    private int getTop(int y) {
        return UIConstants.BorderWidth * (y + 1) + y * this.squareHeight;
    }

    public int getOrder() {
        return this.order;
    }

    /**
     * Paints the cage text for the given square.
     *
     * @param canvas   The canvas to paint on to.
     * @param cageText The cage text to paint on the canvas.
     * @param x        The x index of the square that is painting.
     * @param y        The y index of the square that is painting.
     */
    public void PaintCageText(Canvas canvas, String cageText, int x, int y) {
        int left = this.getLeft(x) + SquareDrawingDimensions.cageTextLeftMargin;
        int top = this.getTop(y) + this.cageTextFontSize;

        canvas.drawText(cageText, left, top, this.cageTextPaint);
    }

    /**
     * Paints the value text for the given square.
     *
     * @param canvas    The canvas to paint on to.
     * @param valueText The value text to paint on the canvas.
     * @param x         The x index of the square that is painting.
     * @param y         The y index of the square that is painting.
     */
    public void PaintValueText(Canvas canvas, String valueText, int x, int y) {

        int textWidth = (int) this.valueTextPaint.measureText(valueText);

        int left = this.getLeft(x) + (this.squareWidth - textWidth) / 2;
        int top = this.getTop(y) + this.cageTextFontSize + this.valueTextFontSize;

        canvas.drawText(valueText, left, top, this.valueTextPaint);
    }

    /**
     * Paints the candidate text for the given square.
     *
     * @param canvas         The canvas to paint on to.
     * @param candidatesText The candidates text to paint on the canvas.
     * @param x              The x index of the square that is painting.
     * @param y              The y index of the square that is painting.
     */
    public void PaintCandidatesText(Canvas canvas, String candidatesText, int x, int y) {
        // Figure out the length of string to print
        int textLength = candidatesText.length();

        // Align the candidates text at the bottom of the square
        int left = this.getLeft(x) + 3;
        int top = this.getTop(y) + this.squareHeight - 5
                - (this.order / 6) * this.candidatesTextFontSize;

        // Note: this is wrapping character by character, not by word
        // first represents the first character index in the string segment
        // line is the line that the text is being written to
        int first = 0;
        int line = 0;
        while (first < textLength) {
            int last = candidatesText.length();
            while (
                    this.candidatesTextPaint.measureText(candidatesText, first, last)
                            > this.squareWidth - 5) {
                last -= 1;
            }

            // Draw the string from first to last
            canvas.drawText(
                    candidatesText,
                    first,
                    last,
                    left,
                    top + line * this.candidatesTextFontSize,
                    this.candidatesTextPaint
            );

            // assign last into first to check for the next substring
            first = last;
            if (first < textLength && candidatesText.charAt(first) == ' ') {
                first += 1;
            }

            line += 1;
        }
    }

    /**
     * Paints the background for the given square.
     *
     * @param canvas The canvas to paint on to.
     * @param paint  The paint to use to draw the background.
     * @param x      The x index of the square that is painting.
     * @param y      The y index of the square that is painting.
     */
    public void PaintBackgroundColor(Canvas canvas, Paint paint, int x, int y) {
        int left = this.getLeft(x);
        int right = left + this.squareWidth;
        int top = this.getTop(y);
        int bottom = top + this.squareHeight;

        canvas.drawRect(
                left,
                top,
                right,
                bottom,
                paint
        );
    }

    public SquareDrawingDimensions(
            int order,
            int squareWidth,
            int squareHeight) {

        this.order = order;
        this.squareWidth = squareWidth;
        this.squareHeight = squareHeight;

        this.cageTextFontSize =
                SquareDrawingDimensions.getCageTextFontSize(this.order);
        this.cageTextPaint = new Paint();
        this.cageTextPaint.setTextSize(this.cageTextFontSize);

        this.valueTextFontSize =
                this.squareHeight
                        - this.cageTextFontSize * 2
                        - 2 * (UIConstants.MaxGameSize - order);
        this.valueTextPaint = new Paint();
        this.valueTextPaint.setTextSize(this.valueTextFontSize);

        this.candidatesTextFontSize = this.squareHeight - this.cageTextFontSize;
        this.candidatesTextPaint = new Paint();
        this.candidatesTextPaint.setTextSize(this.candidatesTextFontSize);
        this.candidatesTextPaint.setColor(UIConstants.CandidatesTextColor);
        String testMeasureString =
                SquareDrawingDimensions.getTestCandidateString(order);

        int maxWidth = this.squareWidth - 5 - (UIConstants.MaxGameSize - order) * 2;

        while (this.candidatesTextPaint.measureText(testMeasureString) > maxWidth) {
            this.candidatesTextFontSize -= 1;
            this.candidatesTextPaint.setTextSize(this.candidatesTextFontSize);
        }
    }
}

File: app/src/main/java/com/anthonysottile/kenken/ui/SquareDrawingDimensions.kt
package com.anthonysottile.kenken.ui

import android.graphics.Canvas
import android.graphics.Paint
import kotlin.math.ceil

class SquareDrawingDimensions(
        val order: Int,
        private val squareWidth: Int,
        private val squareHeight: Int
) {

    private val cageTextFontSize: Int = squareHeight / 5
    private val valueTextFontSize: Int
    private var candidatesTextFontSize: Int = 0

    private val cageTextPaint = Paint()
    private val valueTextPaint = Paint()
    private val candidatesTextPaint = Paint()


    private fun getLeft(x: Int): Int {
        return UIConstants.BorderWidth * (x + 1) + x * this.squareWidth
    }

    private fun getTop(y: Int): Int {
        return UIConstants.BorderWidth * (y + 1) + y * this.squareHeight
    }

    fun paintCageText(canvas: Canvas, cageText: String, x: Int, y: Int) {
        val left = this.getLeft(x) + SquareDrawingDimensions.cageTextLeftMargin
        val top = this.getTop(y) + this.cageTextFontSize

        canvas.drawText(cageText, left.toFloat(), top.toFloat(), this.cageTextPaint)
    }

    fun paintValueText(canvas: Canvas, valueText: String, x: Int, y: Int) {
        val textWidth = this.valueTextPaint.measureText(valueText).toInt()

        val left = this.getLeft(x) + (this.squareWidth - textWidth) / 2
        val top = this.getTop(y) + this.cageTextFontSize + this.valueTextFontSize

        canvas.drawText(valueText, left.toFloat(), top.toFloat(), this.valueTextPaint)
    }

    fun paintCandidatesText(canvas: Canvas, candidatesText: String, x: Int, y: Int) {
        // Figure out the length of string to print
        val textLength = candidatesText.length

        // Align the candidates text at the bottom of the square
        val left = this.getLeft(x) + 3
        val top = (this.getTop(y) + this.squareHeight - 5
                - this.order / 6 * this.candidatesTextFontSize)

        // Note: this is wrapping character by character, not by word
        // first represents the first character index in the string segment
        // line is the line that the text is being written to
        var first = 0
        var line = 0
        while (first < textLength) {
            var last = candidatesText.length
            while (this.candidatesTextPaint.measureText(candidatesText, first, last) > this.squareWidth - 5) {
                last -= 1
            }

            // Draw the string from first to last
            canvas.drawText(
                    candidatesText,
                    first,
                    last,
                    left.toFloat(),
                    (top + line * this.candidatesTextFontSize).toFloat(),
                    this.candidatesTextPaint
            )

            // assign last into first to check for the next substring
            first = last
            if (first < textLength && candidatesText[first] == ' ') {
                first += 1
            }

            line += 1
        }
    }

    fun paintBackgroundColor(canvas: Canvas, paint: Paint, x: Int, y: Int) {
        val left = this.getLeft(x)
        val right = left + this.squareWidth
        val top = this.getTop(y)
        val bottom = top + this.squareHeight

        canvas.drawRect(
                left.toFloat(),
                top.toFloat(),
                right.toFloat(),
                bottom.toFloat(),
                paint
        )
    }

    init {
        this.cageTextPaint.textSize = this.cageTextFontSize.toFloat()

        this.valueTextFontSize = (
                this.squareHeight -
                        2 * this.cageTextFontSize -
                        (UIConstants.MaxGameSize - order)
                )
        this.valueTextPaint.textSize = this.valueTextFontSize.toFloat()

        this.candidatesTextFontSize = this.squareHeight - this.cageTextFontSize
        this.candidatesTextPaint.textSize = this.candidatesTextFontSize.toFloat()
        this.candidatesTextPaint.color = UIConstants.CandidatesTextColor
        val testMeasureString = SquareDrawingDimensions.getTestCandidateString(order)

        val maxWidth = this.squareWidth - 5 - (UIConstants.MaxGameSize - order) * 2

        while (this.candidatesTextPaint.measureText(testMeasureString) > maxWidth) {
            this.candidatesTextFontSize -= 1
            this.candidatesTextPaint.textSize = this.candidatesTextFontSize.toFloat()
        }
    }

    companion object {
        private const val cageTextLeftMargin = 5

        private fun getTestCandidateString(order: Int): String {
            // If the order is >= 6 we have two lines of numbers
            val numbers = if (order >= 6) {
                ceil(1.0 * order / 2).toInt()
            } else {
                order
            }
            return IntArray(numbers).joinToString(" ")
        }
    }
}
